java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -127,147 +133,182 @@ public void execute(EventMessage msg) {
      *                              If left empty, activation key from grains will be used.
      * @param proxyId the proxy to which the minion connects, if any
      */
-    public void registerSSHMinion(String minionId, Optional<Long> proxyId,
-                                  Optional<String> activationKeyOverride) {
-        registerMinion(minionId, true, proxyId, activationKeyOverride);
+    public void registerSSHMinion(String minionId, Optional<Long> proxyId, Optional<String> activationKeyOverride) {
+        Optional<MinionStartupGrains> startupGrainsOpt = SALT_SERVICE.getGrains(minionId,
+                new TypeToken<MinionStartupGrains>() { }, "machine_id", "saltboot_initrd", "susemanager");
+        registerMinion(minionId, true, proxyId, activationKeyOverride, startupGrainsOpt);
+    }
+    /**
+     * Performs minion registration or reactivation..
+     * @param minionId minion id
+     * @param isSaltSSH true if a salt-ssh system is bootstrapped
+     * @param activationKeyOverride label of activation key to be applied to the system.
+     *                       If left empty, activation key from grains will be used.
+     * @param startupGrains Grains needed for initial phase of registration
+     */
+    private void registerMinion(String minionId, boolean isSaltSSH, Optional<Long> proxyId,
+                                Optional<String> activationKeyOverride, Optional<MinionStartupGrains> startupGrains) {
+        Opt.consume(startupGrains,
+            ()-> LOG.error("Aborting: needed grains are not found for minion: " + minionId),
+            grains-> {
+                boolean saltbootInitrd = grains.getSaltbootInitrd();
+                Optional<String> mkey = grains.getSuseManagerGrain().flatMap(sm -> sm.getManagementKey());
+                Optional<ActivationKey> activationKey =
+                        mkey.flatMap(mk -> ofNullable(ActivationKeyFactory.lookupByKey(mk)));
+                Optional<String> validReactivationKey =
+                        mkey.flatMap(mk -> isValidReactivationKey(activationKey, minionId) ? mkey : empty());
```

Comment:
Style nitpick: `Optional` supports `filter`

Reply:
Changed it.  :+1: